### PR TITLE
`megatron.core.datasets` SFTDataset and `get_batch*` functions refactor

### PR DESCRIPTION
**Pull request by asolergi-nv**
# What does this PR do ?

In this PR I introduce a new SFTDataset for text that will live in megatron.core.datasets. In addition, I also include a refactor of various get_batch methods and clear guidance about the expected inputs for the three different regimes (pretraining, SFT, and DynamicCP).
 - SFTDataset (@NVIDIA/datasets @duncanriach @arendu):
   - Based on file prefixes, just like in pretraining. We will generate them in the same way using the same tools/preprocess_data.py script, **but it is important to note that the user will be responsible for adding the chat template to the raw JSONL file data.** We can debate whether it is necessary to provide a basic python script that applies the chat template using the nemotron tokenizer or not.
   - Supports training with MULTIPLE file prefixes using the same code & logic as we do for pretraining with the BlendedMegatronDatasetBuilder
   - We have three indexes (document, sample, and shuffle index) that are memory mapped as in pretraining
     - So we no longer have to load a enormous jsonl file into memory
   - We support training only on assistant answers, as well as toggles to choose whether we want to train on tool calls and reasoning traces.
     - To extract segments of each type (system, user, assistant, tool call, tool answer, reasoning), we need to specify the chat template tokens that delimit each segment. In this PR I include the ones used in the Nemotron3 chat template (Nemotron3ChatTemplateConfig), which I manually verified using the inspect_sft_file_prefix.py script (which I will later remove).
   - We pack sequences using the Modified First-Fit Decreasing algorithm.
   - Moved all the padding / truncation / Constraints to the preprocess_sft_batch method (Next section)
   - Created test_sft_dataset.py that checks functionality.
   - Added SFTDataset documentation
 - Created preprocess_sft_batch method.
   - The main challenge with SFT data is handling the different metadata tensors used in attention computation and the various shape constraints that exist.
   - Fundamentally, we do the following:
     - Pad individual sub-sequences for CP divisibility via pad_thd_sequences_for_cp from TE (when cp_size > 1), producing cu_seqlens_padded
     - Pad or truncate the packed sample to exactly max_seq_len tokens
     - Create a loss mask that zeros out padding tokens and prompt tokens
     - Compute per-segment position IDs from cumulative sequence lengths
     - Compute max_seqlen (the length of the longest sub-sequence)    
   - This method could be implemented as a DataLoader collator, but I decided to place it here explicitly to make it clearer that this preprocessing step is essential.
 - Updated mtp_on_this_rank to use explicit arguments (@NVIDIA/gpt  & @NVIDIA/hybrid-mamba ).
 - Refactored get_batch, get_batch_on_this_cp_rank, and get_batch_on_this_tp_rank functions to use explicit arguments (@NVIDIA/multi-modal @NVIDIA/core-nemo @NVIDIA/datasets @NVIDIA/core-adlr @xiaoyao0115 @parthmannan).
   - I refactored these methods to use explicit arguments so they can be reused across other projects and added tests to ensure that any future changes either add inputs properly or do not break model inputs.
   - Also refactored the pretrain_mamba.py and pretrain_gpt.py scripts so they support the same dataset-related features.
   - Added some tests to the get_batch methods to ensure future changes do not break this function.
 - Deleted megatron/training/datasets/sft_dataset.py (@NVIDIA/datasets)
   - Note that IGNORE_INDEX is already defined in megatron/core/tokenizers/text/libraries/sft_tokenizer.py.
 - Added add_special_tokens option to MegatronTokenizerText (@NVIDIA/tokenizers).
   - This is required to obtain the chat template tokens in order to mask out chat template, reasoning, and tool-calling tokens. It is ONLY used for SFT with SFTDataset and SFTTokenizer; the rest of the tokenizers ignore this argument (keeping exactly the same behavior as before this change).
   - NullTokenizer now takes a string and converts it to ASCII values. Previously, it accepted a space-separated list of values, and the tokenized sample was those integers cast directly. In short, it previously only supported string values such as "32 456 123 542"; now it supports ANY string value.

TODO:
- [ ] Remove `inspect_sft_file_prefix.py`
- [ ] Re-implement changes from https://github.com/NVIDIA/Megatron-LM/pull/3564

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by asolergi-nv** [comment]
/ok to test [4ad9e80](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/4ad9e804687fba2d74cccdd6df57e601de573c6b)

**Comment by asolergi-nv** [comment]
/ok to test [e449719](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/e4497199de98cfbc083723a681a9db791bb3ef45)

**Comment by asolergi-nv** [comment]
/ok to test [01481d3](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/01481d34cd11d1bb72577e983c218e57119b6541)

**Comment by asolergi-nv** [comment]
/ok to test [c832039](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/c832039c59d16ab9fb32148c3922c9ef8886b27d)

**Comment by asolergi-nv** [comment]
/ok to test [12ec5e6](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/12ec5e6d2f6552eb3a0caf998d4d7be9dff6d4ef)

**Comment by asolergi-nv** [comment]
/ok to test [be6f0c8](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/be6f0c80f281d93778351fe77763f6c1d753118e)

**Comment by asolergi-nv** [comment]
/ok to test [4dddd94](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/4dddd94f5713731244cca35a771573cea4d13664)

**Comment by asolergi-nv** [comment]
/ok to test [7740a2c](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/7740a2c6a4b08d1585b23016a753143cfa9de260)

**Comment by asolergi-nv** [comment]
/ok to test [e785d9b](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/e785d9b635ff330652ddccf2db989bff67a517b0)

**Comment by asolergi-nv** [comment]
/ok to test [b9fb498](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/b9fb49870888a3406c4cda0071054f65b8b362b4)

**Comment by asolergi-nv** [comment]
/ok to test [33d9372](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/33d93721f6c9f284fed9bcb175fc5546ce202cd6)

**Comment by asolergi-nv** [comment]
/ok to test [e969aa1](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/e969aa14ba7a6183820045df92c2bdf5e5e2f210)

**Comment by asolergi-nv** [comment]
/ok to test [3e1ddf7](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/3e1ddf71e19b6ebd6e4eda9f7b82ac90c335e4de)

**Comment by asolergi-nv** [comment]
/ok to test [754aa2d](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/754aa2d07faaa850e4d4a62d1d82d8201968c2d2)

**Comment by asolergi-nv** [comment]
/ok to test [d6c9b29](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/d6c9b294f956028025c6651ab09c9beb8171af73)

**Comment by asolergi-nv** [comment]
/ok to test [f07ae4b](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/f07ae4b4064606c129b4c8a3d4a727b963a27f42)

**Comment by asolergi-nv** [comment]
/ok to test [b6e48b1](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/b6e48b1589f1fa4205181ea918c917a3d47df0c9)

**Comment by asolergi-nv** [comment]
/ok to test [c093244](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/c093244e5f38c1de1922c7c6f6d2be6fa6b38d55)

**Comment by asolergi-nv** [comment]
/ok to test [807ee2e](https://github.com/NVIDIA/Megatron-LM/pull/3542/commits/807ee2e06b749c480f7e5bf0676911036434f48a)

**Comment by asolergi-nv** [comment]
We are dividing this PR into 2 smaller PRs. 
First #4103
